_posts/2021-02-02-enso-prediction/enso-prediction.Rmd

```diff
@@ -0,0 +1,2410 @@
+
+After the custom dataset, we create the -- pretty typical -- `dataloader`s, making use of a batch size of 4.
+
+```{r}
+batch_size <- 4
+
+train_dl <- train_ds %>% dataloader(batch_size = batch_size, shuffle = TRUE)
+
+valid_dl <- valid_ds %>% dataloader(batch_size = batch_size)
+
+```
+
+Next, we proceed to model creation.
+
+# Model
+
+The model's main ingredient is the convLSTM introduced in a [prior post](https://blogs.rstudio.com/ai/posts/2020-12-17-torch-convlstm/). For convenience, we reproduce the code in the appendix.
+
+Besides the convLSTM, the model makes use of three convolutional layers, a batchnorm layer and five linear layers. The logic is the following.
+
+First, the convLSTM job is to predict the next month's sea surface temperatures on the spatial grid. For that, we *almost* just return its final state, - almost: We use `self$conv1` to reduce the number channels to one.
+
+For predicting index and phase, we then need to flatten the grid, as we require a single value each. This is where the additional conv layers come in. We *do* hope they'll aid in learning, but we also want to reduce the number of parameters a bit, downsizing the grid (`strides = 2` and `strides = 3`, resp.) a bit before the upcoming `torch_flatten()`.
+
+Once we have a flat structure, learning is shared between the tasks of index and phase prediction (`self$linear`), until finally their paths split (`self$cont` and `self$cat`, resp.), and they return their separate outputs.
+
+(The batchnorm? I'll comment on that in the [Discussion].)
+
+```{r}
+model <- nn_module(
+  
+  initialize = function(channels_in,
+                        convlstm_hidden,
+                        convlstm_kernel,
+                        convlstm_layers) {
+    
+    self$n_layers <- convlstm_layers
+    
+    self$convlstm <- convlstm(
+      input_dim = channels_in,
+      hidden_dims = convlstm_hidden,
+      kernel_sizes = convlstm_kernel,
+      n_layers = convlstm_layers
+    )
+    
+    self$conv1 <-
+      nn_conv2d(
+        in_channels = 32,
+        out_channels = 1,
+        kernel_size = 5,
+        padding = 2
+      )
+    
+    self$conv2 <-
+      nn_conv2d(
+        in_channels = 32,
+        out_channels = 32,
+        kernel_size = 5,
+        stride = 2
+      )
+    
+    self$conv3 <-
+      nn_conv2d(
+        in_channels = 32,
+        out_channels = 32,
+        kernel_size = 5,
+        stride = 3
+      )
+    
+    self$linear <- nn_linear(33408, 64)
+    
+    self$b1 <- nn_batch_norm1d(num_features = 64)
+        
+    self$cont <- nn_linear(64, 128)
+    self$cat <- nn_linear(64, 128)
+    
+    self$cont_output <- nn_linear(128, 1)
+    self$cat_output <- nn_linear(128, 3)
+    
+  },
+  
+  forward = function(x) {
+    
+    ret <- self$convlstm(x)
+    layer_last_states <- ret[[2]]
+    last_hidden <- layer_last_states[[self$n_layers]][[1]]
+    
+    next_sst <- self$conv1(last_hidden)
+    
+    c2 <- last_hidden %>% self$conv2() 
```

Comment:
I think I would keep a single pattern here. ie. choose whether to use `x %>% self$layer`  or `self$layer(x)` but mixing both might confuse some users.

Reply:
that's true! all pipes then ;-)